File: openFaces/source/org/openfaces/component/chart/ChartPopup.java
/*
 * OpenFaces - JSF Component Library 3.0
 * Copyright (C) 2007-2012, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */
package org.openfaces.component.chart;

import org.openfaces.component.LoadingMode;
import org.openfaces.component.window.PopupLayer;
import org.openfaces.util.ValueBindings;

import javax.faces.application.ResourceDependencies;
import javax.faces.application.ResourceDependency;
import javax.faces.context.FacesContext;

@ResourceDependencies({
        @ResourceDependency(name = "jsf.js", library = "javax.faces"),
        @ResourceDependency(name = "default.css", library = "openfaces")
})
public class ChartPopup extends PopupLayer {
    public static final String COMPONENT_TYPE = "org.openfaces.ChartPopup";
    public static final String COMPONENT_FAMILY = "org.openfaces.ChartPopup";
    private LoadingMode loadingMode;

    public ChartPopup() {
        setRendererType("org.openfaces.ChartPopupRenderer");
    }

    @Override
    public String getFamily() {
        return COMPONENT_FAMILY;
    }

    @Override
    public String getClientId(FacesContext context) {
        return super.getClientId(context);
    }

    @Override
    public Object saveState(FacesContext context) {
        return new Object[]{
                super.saveState(context),
                saveAttachedState(context, loadingMode)
        };
    }

    @Override
    public void restoreState(FacesContext context, Object stateObj) {
        Object[] state = (Object[]) stateObj;
        int i = 0;
        super.restoreState(context, state[i++]);
        loadingMode = (LoadingMode) restoreAttachedState(context, state[i++]);
    }

    public LoadingMode getLoadingMode() {
        return ValueBindings.get(this, "loadingMode", loadingMode, LoadingMode.CLIENT, LoadingMode.class);
    }

    public void setLoadingMode(LoadingMode loadingMode) {
        this.loadingMode = loadingMode;
    }
}

File: openFaces/source/META-INF/resources/openfaces/3_0/chart/chartPopup.js
/*
 * OpenFaces - JSF Component Library 2.0
 * Copyright (C) 2007-2010, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */
// -------------------------- CHART MENU SUPPORT
O$.ChartPopup = {
  _init: function(chartPopupId, loadingMode, chartId) {
    var chartPopup = O$.initComponent(chartPopupId, null, {
      _loadingMode: loadingMode,
      _contentLoaded: loadingMode == "client",
      _hoverStateHandler: function(mouseInside, element) {
        element._mouseInside = mouseInside;

        setTimeout(function () {
          if ((chartPopup._targetArea && !chartPopup._targetArea._mouseInside) && !chartPopup._mouseInside) {
            chartPopup.hide();
          }
        }, 100);
      }
    });

    O$.setupHoverStateFunction(chartPopup, chartPopup._hoverStateHandler);
  },

  _ajaxResponseProcessor: function(popup, portionName, portionHTML, portionScripts) {
    var oldComponent, prnt, tempDiv, newPopup, oldId;
    if (portionName == "content") {
      oldComponent = O$(popup.id);
      prnt = oldComponent.parentNode;
      tempDiv = document.createElement("div");
      tempDiv.innerHTML = portionHTML;
      newPopup = tempDiv.childNodes[0];
      oldId = oldComponent.id;
      prnt.replaceChild(newPopup, oldComponent);
      newPopup.id = oldId;
    }
    O$.executeScripts(portionScripts);
    newPopup._contentLoaded = true;
    newPopup._targetArea = popup._targetArea;
    var chart = newPopup.parentNode;
    if (chart._visiblePopup) {
      chart._visiblePopup.hide();
    }
    newPopup.showAtXY(popup._eventX + 2, popup._eventY + 2);
    chart._visiblePopup = newPopup;
  },

  show:function(event, chartPopupId) {
    var popup = O$(chartPopupId);
    var chart = popup.parentNode;
    var eventPoint = O$.getEventPoint(event);
    var targetArea = event.target ? event.target : event.srcElement;
    popup._targetArea = targetArea;

    if (!targetArea._initialized) {
      O$.setupHoverStateFunction(targetArea, popup._hoverStateHandler);
      targetArea._initialized = true;
    }

    if (popup._loadingMode == "ajaxAlways" || !popup._contentLoaded) {
      popup._eventX = eventPoint.x;
      popup._eventY = eventPoint.y;
      O$.requestComponentPortions(popup.id, ["content"], null, O$.ChartPopup._ajaxResponseProcessor);
    } else {
      if (!popup.isVisible()) {
        if (chart._visiblePopup) {
          chart._visiblePopup.hide();
        }
        popup.showAtXY(eventPoint.x + 2, eventPoint.y + 2);
        chart._visiblePopup = popup;
      }
    }
  },

  hide:function(event, chartPopupId) {
    var popup = O$(chartPopupId);

  }
};


